### PR TITLE
nixos/fanout: init fanout oneshot module

### PR DESCRIPTION
**Pull request by therishidesai**
###### Description of changes

Added a oneshot daemon to setup the [fanout](https://github.com/bob-linuxtoys/fanout) kernel module.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/config/fanout.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    boot.kernelModules = [ "fanout" ];
+
+    boot.extraModprobeConfig = ''
+      options fanout buffersize=${builtins.toString cfg.bufferSize}
+    '';
+
+    systemd.services.fanout = {
+      description = "Bring up /dev/fanout devices";
+      script = ''
+        MAJOR=`${pkgs.gnugrep}/bin/grep fanout /proc/devices | ${pkgs.gawk}/bin/awk '{print $1}'`
```
Comment (SuperSandro2000) [inline nixos/modules/config/fanout.nix:33]:
```suggestion
        MAJOR=$(${pkgs.gnugrep}/bin/grep fanout /proc/devices | ${pkgs.gawk}/bin/awk '{print $1}')
```

File: nixos/modules/config/fanout.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        for ((i = 0; i < ${builtins.toString cfg.fanoutDevices}; i++))
+        do
+          mknod /dev/fanout$i c $MAJOR $i
+        done
```
Comment (SuperSandro2000) [inline nixos/modules/config/fanout.nix:37]: We can generate the commands in nix

File: nixos/modules/config/fanout.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.fanout;
+in
+{
+  options.services.fanout = {
+    enable = mkEnableOption "fanout";
```
Comment (SuperSandro2000) [inline nixos/modules/config/fanout.nix:8]:
```suggestion
    enable = mkEnableOption (lib.mdDoc "fanout");
```

File: nixos/modules/config/fanout.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    fanoutDevices = mkOption {
+      description = "Number of /dev/fanout devices";
+      type = types.int;
+      default = 1;
+    };
+    bufferSize = mkOption {
+      description = "Size of /dev/fanout buffer in bytes";
+      type = types.int;
+      default = 16384;
+    };
```
Comment (SuperSandro2000) [inline nixos/modules/config/fanout.nix:18]:
```suggestion
      type = types.int;
      default = 1;
      description = "Number of /dev/fanout devices";
    };
    bufferSize = mkOption {
      type = types.int;
      default = 16384;
      description = "Size of /dev/fanout buffer in bytes";
    };
```

File: nixos/modules/config/fanout.nix [hunk 5]
```diff
@@ -0,0 +1,50 @@
+{ config, lib, pkgs, ... }:
+with lib;
```
Comment (AndersonTorres) [inline nixos/modules/config/fanout.nix:2]:
```suggestion

```

File: nixos/modules/config/fanout.nix [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    };
+    bufferSize = mkOption {
+      type = types.int;
+      default = 16384;
+      description = "Size of /dev/fanout buffer in bytes";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    boot.extraModulePackages = with config.boot.kernelPackages; [ fanout.out ];
```
Comment (AndersonTorres) [inline nixos/modules/config/fanout.nix:25]:
```suggestion
    boot.extraModulePackages = [ config.boot.kernelPackages.fanout.out ];
```

File: nixos/tests/all-tests.nix [hunk 7]
```diff
@@ -262,6 +262,7 @@ in {
   fail2ban = handleTest ./fail2ban.nix { };
   fakeroute = handleTest ./fakeroute.nix {};
   fancontrol = handleTest ./fancontrol.nix {};
+  fanout = handleTest ./fanout.nix {};
```
Comment (AndersonTorres) [inline nixos/tests/all-tests.nix:265]: Split the tests in a separate commit.

File: nixos/tests/fanout.nix [hunk 8]
```diff
@@ -0,0 +1,32 @@
+{ system ? builtins.currentSystem
+, config ? {}
+, pkgs ? import ../.. { inherit system config; }
+}:
```
Comment (AndersonTorres) [inline nixos/tests/fanout.nix:4]:
```suggestion
,  lib
}:
```

File: nixos/tests/fanout.nix [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import ./make-test-python.nix {
+  name = "fanout";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ therishidesai ];
+  };
```
Comment (AndersonTorres) [inline nixos/tests/fanout.nix:9]:
```suggestion
  meta.maintainers = [ lib.maintainers.therishidesai ];
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by AndersonTorres — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by Nyabinary** [comment]
@SuperSandro2000 This PR been sitting for a while, a merge should probably happen unless anything is wrong.
